### PR TITLE
Plan Decorator

### PR DESCRIPTION
**Pull request by cjtitus**
## Summary

This PR would add a plan decorator that can be used to wrap any plan in a simple plan class, that at the minimum, provides a warning when a plan is used without `yield from`. 

## Motivation and Context

Plan decorators have been discussed intermittently as a building-block to enable a few different features. The code this PR is built on was proposed by @danielballan in #1239 as a way to warn users if they have forgotten to `yield from` a plan, and implement a more useful `repr` for plans than the built in `<generator object my_plan at 0x7f3cc009cc80>` you usually see.

Plan decorators would also allow for better plan inspection capabilities, as we could differentiate between a _plan_ and a normal generator. In the future (i.e, not in this PR), different plan decorators could be defined for, e.g, adaptive plans, so that a summarizer could check what _kind_ of plan it is trying to summarize, and act appropriately.

More controversially, #1238 proposed plan decorators as a possible building block towards shortcutting the `RE(...)` syntax. This excites some, and scares others, and will not be explored in this PR.

## Current Implementation
This code is pulled whole-cloth from [Dan's code in #1239](https://github.com/bluesky/bluesky/issues/1239#issue-479177515 ), with only minor adjustments to the `repr` to make it more friendly. 

It does currently:
* Warn users if they neglect to `yield from` a plan in another plan.
* Change the `repr` to be `<bluesky plan: {plan_name}>`

It should (by the time we merge):
* Decorate all bluesky plans

It does not:
* Warn users if they neglect to call `RE(...)` on a plan
* Implement any of the "Plan Accessor" concept discussed in #1239
* Implement a dynamic `repr` that contains information about the generator state, as discussed in #1239

## Please Discuss
The big question is - how much functionality should we give this plan decorator? Right now it is very minimal. It is a building block that implements a useful warning, and will allow _future_ code to reason about whether or not it has a bluesky plan. Plan accessors would be a _major_ shift in how bluesky is able to be used, so I am _very reluctant_ to delve into them here, despite the strong previous support. My preference is to take the code that Dan wrote in 2019, and shepherd it into bluesky so that we can start using it.

However, I would very much welcome all feedback and discussion.

## How Has This Been Tested?

A test has been added to see if an improperly called plan warns correctly, and a properly called plan runs error-free.

### CODE REVIEW

**Comment by cjtitus** [comment]
One interesting side-effect of the implementation via `__del__` is that attempting to introspect plans will actually throw warnings if you try to do `isinstance(test_plan(), Plan)`, you will get the warning about missing `yield from`.

Another interesting side-effect of the wrapping is that `type(test_plan())` is `bluesky.preprocessors.plan`, but `type(test_plan)` is `function`. This isn't any different from how normal generators work, but I do wonder if we could make it less confusing, or somehow indicate that the function will _return_ a generator

**Comment by ksunden** [comment]
I suspect that the `type(test_plan)` behavior will interact negatively with queueserver, which uses `inspect.isgeneratorfunction` checks to find generator functions:

https://github.com/bluesky/bluesky-queueserver/blob/main/bluesky_queueserver/manager/profile_ops.py#L422-L426

**Comment by cjtitus** [comment]
If we _did_ implement a `type(test_plan)`, queueserver would probably have to switch to using that, rather than `inspect.isgeneratorfunction`. It looks like that would be an easy swap. It would also result in queueserver only finding _plans_, rather than any random generator function in the namespace.

Is that something that would be desirable from a bluesky-queueserver perspective? I think that would be better than using `isgeneratorfunction`, no? I was thinking that the desire to get everything that's _actually_ a bluesky plan would probably be a benefit.

**Comment by prjemian** [comment]
So exactly how does/will `bluesky.isbluskyplanfunction(obj)` work?  How does it identify that the function has been decorated without creating the generator?

**Comment by prjemian** [comment]
I'm thinking about the [`__wrapped__`](https://docs.python.org/3/library/functools.html#functools.update_wrapper) attribute in [`functools`](https://docs.python.org/3/library/functools.html).

**Comment by cjtitus** [comment]
Well, one thing you could do is use a class to wrap your generator, rather than a function. So rather than 
```python
def plan(gen_instance):
    @wraps
    def inner_plan(*args, **kwargs):
        return Plan(gen_instance(*args, **kwargs))
    return inner_plan
```

You can actually write
```python
class plan_function:
    def __init__(self, gen_instance):
        self.gen_instance = gen_instance
        functools.update_wrapper(self, gen_instance)
    def __call__(*args, **kwargs):
        return Plan(self.gen_instance(*args, **kwargs))
```

so now when you apply 
```python
@plan_function
def my_plan():
    yield Msg('null')
```
you can do `isinstance(my_plan, plan_function) == True`, and `isinstance(my_plan(), Plan) == True`

Now, most of my knowledge about the details of wrappers has admittedly come in the last 24 hours, courtesy of [how you implemented your python decorator is wrong.](https://github.com/GrahamDumpleton/wrapt/blob/develop/blog/01-how-you-implemented-your-python-decorator-is-wrong.md) So I'm aware that if things could get arbitrarily more complex if, like the author of that blog, you want to be able to wrap arbitrary objects with great robustness. However, I think this class-based wrapper would probably suffice to be able to wrap _generators_ and let us know that they're plans without instantiating them.

**Comment by cjtitus** [comment]
To be even a bit more concise and answer @prjemian directly, the function `bluesky.isblueskyplan(obj)` would be
```python
def isblueskyplanfunction(obj):
    return isinstance(obj, plan_function)
```
And in fact there may not even be a need for the utility function, unless you want to hide the `plan_function` implementation.

**Comment by cjtitus** [comment]
Even more things to like about class-based wrappers: 
```python
class plan_function:
    def __init__(self, gen_instance):
        self.gen_instance = gen_instance
        functools.update_wrapper(self, gen_instance)

    def __call__(self, *args, **kwargs):
        return Plan(self.gen_instance(*args, **kwargs))

    def __repr__(self):
        return f"<Uninitialized bluesky plan: {self.gen_instance.__name__}"
```
is a perfect way to get a sensible `repr` for a bluesky plan that hasn't actually been called. This is something that was definitely requested in #1239. 

This way, `repr(my_plan)` returns `<Uninitialized bluesky plan: my_plan>`, and `repr(my_plan())` returns `<bluesky plan: my_plan>`, and we could make these arbitrarily distinct.

**Comment by cjtitus** [comment]
I have gone and implemented class-based decorators. As @ksunden pointed out, this will break bluesky-queueserver's plan identification for decorated plans. I think @dmgav is probably the correct person to alert. An easy fix would be to use the new `isplanfunction()` to catch new decorated plans, and `isgeneratorfunction()` to catch non-decorated plans. One could go a step further, and have a "strict" argument to `isplanfunction()` that would either include or exclude non-decorated plans.

**Comment by dmgav** [comment]
My concern is how well access to the signature and docstring of the original plan will work. It needs to be tested before this PR is merged.

**Comment by cjtitus** [comment]
Docstring and signature work out-of-the-box, I assume due to functools.update_wrapper. For signature, you need inspect.signature, inspect.getfullargspec returns the signature of the actual class __call__ function, which I assume is intentional.

I can add tests to make sure that these work.

**Comment by dmgav** [comment]
I tested how `inspect.signature()` with the class instance. It seems to work fine.

**Comment by cjtitus** [comment]
>     * Since `PlanFunction` is a class, it may be useful to override `__getattr__` so that we could access attributes of the original plan function.

What attributes would the original plan function have?

**Comment by cjtitus** [comment]
We don't need to override `__getattr__`, because `update_wrapper` already copies the `__dict__` over from the function you are wrapping. So doing the following:
```python
def test_plan():
    yield Msg("null")

test_plan.test_attr = "a test"
test_plan2 = PlanFunction(test_plan)
assert test_plan2.test_attr == "a test"
```
works as you would expect (or if you're a pessimist, it works better than you expect!)

**Comment by dmgav** [comment]
@cjtitus Thanks, you are right. It is unlikely that we may need to set attributes of wrapped functions or monitor dynamically changed attributes, so copy of the `__dict__` should be sufficient.

**Comment by cjtitus** [comment]
It is a good point that right now this doesn't follow the naming convention for decorators. I would like a more intuitive wrapper name than "PlanFunction". `@plan` would be nice, but it's also not good to have the only distinction between two very related classes be the capitalization of the first letter.

maybe the two classes could be called `plan_decorator` and `inner_plan`, and `plan_decorator` is the one that is actually used to decorate things. Probably people should never have to check these types themselves, as we can provide  `isplan(obj)` and `isplanfunction(obj)` functions, so that nobody needs to worry about what we called the innermost plan decorator class.

**Comment by dmgav** [comment]
I think the name of the `Plan` class is good. The name `plan_decorator` follows naming convention.

**Comment by cjtitus** [comment]
Yeah, I'm just worried that someone might accidentally try to wrap a function with `Plan` directly, rather than `plan_decorator`. It would be better to somehow make it clear that `Plan` is an internal class that should never be called on its own.

**Comment by dmgav** [comment]
We could name the class as `_Plan`. This should discourage using this class outside Bluesky package. It may be also useful to have `plan_wrapper` that creates `_Plan` class from a generator:
```
def plan_wrapper(gen_instance):
    return _Plan(gen_instance)
```
Then documented public API will include `plan_wrapper`, `plan_decorator`, `isplan`, `isplanfunction`. 
@tacaswell, what is your opinion?

**Comment by cjtitus** [comment]
Gotta admit, I don't really understand the distinction that is being made between a wrapper and a decorator. The distinction that seems to be made in Bluesky is that "wrappers" don't have any of the `@wraps` machinery that updates the name, docstrings, etc, and `make_decorator` adds that in. This is a pretty technical distinction that most users probably don't know about.

Do we want to have the ability to wrap a plan in the `_Plan` class and _not_ update the name, docstring, or signature? It seems like that could be confusing.

**Comment by dmgav** [comment]
In bluesky project the difference between wrapper and decorator is that wrappers accepts generators and decorators accepts functions as the parameters: https://github.com/bluesky/bluesky/blob/0f19587dbaf3d70029c77f63a04d39f794ae1806/bluesky/utils/__init__.py#L1121-L1128 Sometimes wrappers are more convenient.

**Comment by cjtitus** [comment]
Names changed to be consistent with decorator/wrapper nomenclature.

The one thing that still bugs me is that the `__del__` based warning system will also warn you if you do something like check
`isplan(my_plan())`, because `my_plan()` gets instantiated, checked-but-not-pulled, and then destroyed, triggering the warning message. In theory, people shouldn't need to do this, because they can check `isplanfunction(myplan)`, but it could still be confusing.

I do not have any ideas about how to get around this, and I think it's 100% worth it to have a warning that a plan was not yielded from.

**Comment by cjtitus** [comment]
Additionally, @danielballan @tacaswell please advise on whether or not this PR should attempt to actually decorate all of the built-in plans and plan stubs. I won't go decorating things until it's been reviewed though, in case the name or implementation changes.

**Comment by ksunden** [comment]
One point to consider is how does this plan decorator interact with the other decorators in the ecosystem (especially preprocessors)?

e.g. if I want to do `modified_grid_scan = baseline_decorator([foo, bar])(grid_scan)`, would the modified version be identified properly by `isplanfunction` etc?

**Comment by cjtitus** [comment]
This is a good point. The modified version is _not_ identified correctly by `isplanfunction`, because plan wrappers and `make_decorator` return generator functions, not `plan_decorator` class instances. It also wipes out the warning when you don't pull from the generators.

This could possibly be "fixed" by having `make_decorator` return a `plan_decorator`, essentially making any function that uses a decorator into a plan. I do worry a bit that this would result in one layer of `plan_decorator` objects for each decorator that you use. Unfortunately I don't know of a different way to do it.

File: bluesky/tests/test_generators.py [hunk 1]
```diff
@@ -462,3 +463,47 @@ def plan():
...
+    EchoRE(correct_plan())
+
+
+def test_is_a_plan_function():
+
+    @plan_decorator
+    def test_plan():
+        yield Msg("null")
+
+    def not_a_real_plan():
```
[the remark this answers is not in the document]
Comment (cjtitus) [reply bluesky/tests/test_generators.py:498]: This is definitely a fair criticism. Even in the tests we convey intent, so I can switch to "not_a_decorated_plan".

This suggests another question -- do we want the plan-decorator-tester functions to be called `isplanfunction` and `isplan`? Or should we make them something like `isdecoratedplan` instead? That would make it more clear what the function means, and avoid confusion if people wonder why their un-decorated plans (which are clearly plans) don't get recognized as plans.

**Comment by prjemian** [comment]
Just as there is a difference between generators and generator functions,
Tom has pointed out that a bluesky plan can be a list of Msg objects. We
need to make that distinction here.

isplanfunction() gets my vote

**Comment by cjtitus** [comment]
Ok, so the current version has switched to checking "_plan" and "_plan_function" attributes, so "isplanfunction" at least will keep working even if you decorate over the `@plan_decorator`. However, the missing `yield from` warnings won't work, and neither will `isplan` if you decorate or wrap on top of the `@plan_decorator`.

These last two are fundamental issues that I don't see any resolution to unless somebody comes up with a brilliant way of doing warnings without relying on __del__, or we modify all of the wrappers and `make_decorator` function to essentially redecorate everything with `@plan_decorator`.

So the state is that `@plan_decorator` works quite well if it is the last thing to decorate a plan, but the functionality is very degraded if it is _not_ the last thing to decorate a plan. I was sort of hoping that I would be able to preserve more functionality than that. As the functionality is currently pretty bare-bones, I'm not sure it's worth the potential confusion. But if `@plan_decorator` became a base to implement other functionality, maybe it is worth it.

Opinions?

**Comment by stan-dot** [comment]
@cjtitus  at the moment we have a tracing decorator here: https://github.com/bluesky/bluesky/blob/main/src/bluesky/tracing.py

would this fill your use case?

**Comment by cjtitus** [comment]
This concept is something that, at the moment, I have entirely abandoned. I think that if it had been built in to the concept of Bluesky from the start it could have been useful, but in the present ecosystem, getting a sensible amount of functionality would require modifying too many of the built-in plans and decorators, especially when QueueServer is considered.

Oddly, moving to a GUI side-steps the need for much of this functionality, as the Queue replaces all of the "compositional" plans I was writing, getting a good `repr` is much less important, and most plans are expected to have a widget that builds the arguments and adds them to the queue.

I'll close this to clean up the PR list.

**Pull request closed without merging**